README.md

# ps1-4git

#### A git-aware section for your PS1 prompt

## What is it?

I wanted a git-aware prompt and was not able to get other things working
the way I wanted (my fault, definitely). On the other hand I wanted to
remind myself C coding (didn't do that since ~ 1993) so I decided to
combine these two into one -- code in C a super simple and lightweight
git-aware prompt add-on for my bash ``PS1``! ;)


## How does it look like?

(Working on providing screenshots, stay tuned.)


## Prerequisites

To be able to build the ``ps1-4git`` binary, you need a C compiler as well as
``make`` installed on your system.

The following will do that for you on Debian/Ubuntu:

```bash
$ sudo apt-get install build-essential make
```

## Build and install

To make use of this utility -- download, build and install:

```bash
$ git clone --depth=1 https://github.com/yassen-itlabs/git-goodies.git
$ cd git-goodies/ps1
$ make && sudo make install
$ source ~/.bashrc
```

**What happened during the install?**

The ``make install`` step should have added to your ``~/.bashrc`` a line
that appends ``ps1-4git`` output to your ``PS1`` prompt, similar to what is shown
below:

```bash
PS1="$PS1\$(git status 2>&1 | ps1-4git -n)"
```

The ``ps1-4git`` binary outputs such a line when called with ``--4bashrc``, like:

```bash
ps1-4git --4bashrc >> ~/.bashrc"  # make sure you do not add this multiple times
```

**What other command line arguments are supported?**

See ``ps1-4git --help``.


## Uninstall

When you decide you want to get rid of ``ps1-4git``, just go to the
``git-goodies/ps1`` directory (or clone it as shown in the _Build
and Install_ section) and execute ``sudo make uninstall``:

```bash
$ cd git-goodies/ps1
$ sudo make uninstall
```

After that, you can safely remove the repo directory itself like:

```bash
cd ~/parent-where-git-goodies-is/
rm -rf git-goodies
```


## Tests

You probably don't need to do this (unless you want to make changes to the code
and possibly submit a pull request). Still, if you need or want to, read further.

To be able to build and run the tests, you need ``CUnit`` development package
installed on your system.

On Debian/Ubuntu this can be done with:

```bash
$ sudo apt-get install libcunit1-dev
```

(tested and working on Ubuntu 18.04).

As soon as you have that, going to the ``ps1/`` directory and running ``make test``
should work:

```bash
$ cd git-goodies/ps1
$ make test
```

Expected output:

```
user@host:~/some-dir/git-goodies/ps1$ make test
gcc  -O2 -march=native ps1-4git.c tests/test-ps1-4git.c -Wall -Wfloat-equal -Wextra  -lcunit -o tests/test-ps1-4git
./tests/run-test.sh

All tests passed - OK
```

## License
Released under the **MIT** license.

Copyright (c) 2019, Yassen Damyanov

This program is free software: you can redistribute it and/or modify
it under the terms of the MIT License.

You should have received a copy of the MIT License along with this program
(see LICENSE text file).

If not, see <https://opensource.org/licenses/MIT>.


## Author

Primary author/maintainer (for now):

Yassen Damyanov <``yassen.damyanov.bg -AT- gmail.com``>


CHANGELOG.md

# git-goodies / ps1-4git CHANGELOG

Ver. 0.2.2
----------
_Released 2019-08-24_ <br>
Fix problems in prompt line due to wrong character count coming from
counting non-printable characters (#1).

Ver. 0.2.1
----------
_Released 2019-08-18_ <br>
Add option to switch off unicode symbols in output.

Ver. 0.1.3
----------
_Released 2019-08-14_ <br>
First version released internally for friends to try.


ps1/tests/run-test.sh
#! /bin/sh

# ps1-4git - A git-aware section for your PS1 prompt
# Copyright (c) 2019, Yassen Damyanov
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License.
# You should have received a copy of the MIT License along with this program.
# If not, see <https://opensource.org/licenses/MIT>.

# Run test-ps1-4git and hide output on success; show on failed tests.
# Assuming execution is done from parent directory ps1/.

_out=$(./tests/test-ps1-4git 2>&1)  #  && _ok=yes || _ok=no
_tests_failed=$(echo "$_out" | awk '/[ ]+tests[ ]+/ {print $5}')

if [ x"$_tests_failed" = x0 ]; then
    echo "\nAll tests passed - OK"
else
    msg="***** $_tests_failed test(s) FAILED!"
    echo "\n\n$msg$_out\n\n$msg"
fi


ps1/tests/test-ps1-4git.c
/*
ps1-4git - A git-aware section for your PS1 prompt
Copyright (c) 2019, Yassen Damyanov

This program is free software: you can redistribute it and/or modify
it under the terms of the MIT License.
You should have received a copy of the MIT License along with this program.
If not, see <https://opensource.org/licenses/MIT>.
*/

/**
 * Unit tests for the ps1-4git module.
 *
 * Helpful URLs:
 *   https://stackoverflow.com/questions/38183974/cunit-assertion-assertion-void-0-f-pcursuite-failed
 *   http://cunit.sourceforge.net/doc/writing_tests.html
 *
 * See README.md on how to build and run the tests.
 *
 *   TODO:
 *    - Cover parse_input()
 *    - Provide end-to-end tests that send the binary different git-status outputs (python-based?)
 */

#include <stdlib.h>

#include <CUnit/CUnit.h>
#include <CUnit/Basic.h>

#include "../ps1-4git.h"


void test_rindexof_expect_7(void) {
    //               0000000000111
    //               0123456789012
    char* somestr = "one two three";
    CU_ASSERT(7 == rindexof(' ', somestr, -1));
    CU_ASSERT(7 == rindexof(' ', somestr, 12));
    CU_ASSERT(7 == rindexof(' ', somestr, 10));
    CU_ASSERT(7 == rindexof(' ', somestr, 7));
}


void test_rindexof_expect_3(void) {
    //               0000000000111
    //               0123456789012
    char* somestr = "one two three";
    CU_ASSERT(3 == rindexof(' ', somestr, 6));
    CU_ASSERT(3 == rindexof(' ', somestr, 4));
    CU_ASSERT(3 == rindexof(' ', somestr, 3));
}


void test_rindexof_crawling_to_the_left(void) {
    //            0000000000111111111122222222223333333333444444444455
    //            0123456789012345678901234567890123456789012345678901
    char* line = "Your branch is ahead of 'origin/master' by 1 commit.";
    int ahead_right = rindexof(' ', line, -1);
    int ahead_left = rindexof(' ', line, ahead_right -1);
    int aheadof_count = atoi(&line[ahead_left + 1]);
    CU_ASSERT(44 == ahead_right);
    CU_ASSERT(42 == ahead_left);
    CU_ASSERT(1 == aheadof_count);
}


void test_startswith_true(void) {
    CU_ASSERT_TRUE(startswith("hello world", "hello"));
    CU_ASSERT_TRUE(startswith("", ""));
    CU_ASSERT_TRUE(startswith("anything", ""));
}


void test_startswith_false(void) {
    CU_ASSERT_FALSE(startswith("hello world", "the"));
    CU_ASSERT_FALSE(startswith("", "a"));
}


void test_isoneof_true(void) {
    CU_ASSERT_TRUE(isoneof("-s", "--sample-switch", "-s"));
    CU_ASSERT_TRUE(isoneof("-s", "--sample-switch", "--sample-switch"));
}


void test_isoneof_false(void) {
    CU_ASSERT_FALSE(isoneof("-s", "--sample-switch", "--sample-"));
    CU_ASSERT_FALSE(isoneof("-s", "--sample-switch", "--sample-switc"));
    CU_ASSERT_FALSE(isoneof("-s", "--sample-switch", ""));
}


int run_test_suites(void) {
    CU_initialize_registry();
    CU_pSuite suite = CU_add_suite("ps1-4git-test", 0, 0);

    CU_add_test(suite, "test_rindexof_expect_7", test_rindexof_expect_7);
    CU_add_test(suite, "test_rindexof_expect_3", test_rindexof_expect_3);
    CU_add_test(suite, "test_rindexof_crawling_to_the_left", test_rindexof_crawling_to_the_left);
    CU_add_test(suite, "test_startswith_true", test_startswith_true);
    CU_add_test(suite, "test_startswith_false", test_startswith_false);
    CU_add_test(suite, "test_isoneof_true", test_isoneof_true);
    CU_add_test(suite, "test_isoneof_false", test_isoneof_false);

    CU_basic_set_mode(CU_BRM_NORMAL); // CU_BRM_NORMAL, CU_BRM_SILENT, CU_BRM_VERBOSE
    CU_basic_run_tests();
    unsigned int tests_failed = CU_get_number_of_tests_failed();
    CU_cleanup_registry();

    int rc = tests_failed? 5: 0;
    return rc;

}


int main(void) {
    return run_test_suites();
}


ps1/git-ps1.c
/**
 * Scan 'git status' output (read from stdin) to detect git branch
 * and status of workspace files.
 */

#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define VERSION "0.1.3"

#define EOL 10
#define SPC 32
#define NUL '\0'
#define COMMA ','

#define PATH_MAX 255
#define LINE_MAX 1024
#define BRANCH_MAX 255

#define COL_RED "\e[31m"
#define COL_GRN "\e[32m"
#define COL_YEL "\e[33m"
#define COL_OFF "\e[0m"

#define PS1_SETUP_STRING "PS1=\"$PS1\\$(git status 2>&1 | ps1-4git)\"  # ps1-4git-MARKER: Do NOT remove marker\n"
#define MSG_WRONG_ARGS "Wrong arguments. Try 'ps1-4git -h'\n"

const char* MSG_USAGE =
    "ps1-4git - fancy git dash(board) in the prompt\n"
    "\n"
    "Usage:\n"
    "\n"
    "  ps1-4git -h|--help          Print this help screen.\n"
    "  ps1-4git -v|--version       Show program version.\n"
    "\n"
    "  git status 2>&1 | ps1-4git [-n|--no-unicode]\n"
    "                                Parse 'git status' output and print informative\n"
    "                                prompt ending string.\n"
    "\n"
    "                             -n|--no-unicode switch causes the output to be\n"
    "                                ascii only (no fancy arrows or bullets)\n"
    "\n"
    "                                Put this in $() at the end of your PS1 like:\n"
    "                                  PS1=\"$PS1\\$(git status 2>&1 | ps1-4git)\"\n"
    "\n"
    "  ps1-4git --4bashrc            Print a PS1 definition like the one above\n"
    "                                suitable for inclusion in .bashrc.\n"
;


/**
 * Return index of given char's occurance (searching left to right)
 * in str if found, -1 otherwise.
 */
int indexof(const char ch, const char *str, int start) {
    int idx = start;
    while (str[idx] != NUL && str[idx] != ch) {
        idx ++;
    }
    return (NUL == str[idx])? -1 : idx;
}


/**
 * Return index of given char's occurance (searching right to left)
 * in str if found, -1 otherwise.
 */
int rindexof(const char ch, const char *str, int start) {
    int idx = (start > -1)? start : strlen(str) - 1;
    while (idx >= 0 && str[idx] != ch) {
        idx --;
    }
    return idx;
}


/**
 * Return true if string a starts with string b, false otherwise.
 */
bool startswith(const char *a, const char *b) {
    return (strncmp(a, b, strlen(b)) == 0);
}


/**
 * Expect the output of 'git status' on stdin; scan it for certain patterns
 * and output appropriate combination of strings with data forming he last
 * part of a PS1 shell prompt. Return appropriate result code.
 * If no_unicode is true -- use ascii characters in place of symbols (arrows,
 * bullets).
 */
int parse_input(bool no_unicode) {

    char line[LINE_MAX];
    char thebranch[BRANCH_MAX];

    bool to_be_committed = false;
    bool not_staged = false;
    bool untracked = false;

    bool uptodate = false;
    int aheadof_origin = 0;
    int aheadof_count = -1;
    int behind_count = -1;

    while (fgets(line, sizeof(line), stdin) != NULL) {  // reads newline too
        if (startswith(line, "fatal: ")) {
            return 128;  // return same rc as 'git status' when not in a git repo

        } else if (startswith(line, "On branch ")) {
            // overwrite line EOL with NUL to get rid of it
            char* p = strchr(line, EOL);
            line[(int)(p - line)] = NUL;  // (p - line) is the index of the EOL char
            // get the last word as the branch name
            strcpy(thebranch, line + sizeof("On branch ") - 1);

        } else if (startswith(line, "Your branch is ahead of ")) {
            int right = rindexof(SPC, line, -1);
            int left = rindexof(SPC, line, right -1);
            aheadof_count = atoi(&line[left + 1]);

        } else if (startswith(line, "Your branch is behind ")) {
            int comma = rindexof(COMMA, line, -1);
            int right = rindexof(SPC, line, comma -1);
            int left = rindexof(SPC, line, right -1);
            behind_count = atoi(&line[left + 1]);

        } else if (startswith(line, "Your branch is up to date ")) {
            uptodate = true;

        } else {
            not_staged = not_staged || startswith(line, "Changes not staged for commit:");
            to_be_committed = to_be_committed || startswith(line, "Changes to be committed:");
            untracked = untracked || startswith(line, "Untracked files:");
        }
    }

    // Unicodes: middle: A537 (used)  big: 27A1+space, super small bulet 2022 or 2981)
    // useful unicode char map: https://unicode-table.com/en/#hangul-syllables

    char* bullet = no_unicode? "*" : "ꔷ";
    char* left_arrow = no_unicode? "<-" : "⬅ ";
    char* right_arrow = no_unicode? "->" : "➡ ";

    printf("(");  // TODO: denote push with ⭡2 (2B61), pull with ⭣3 (2B63)
    if (to_be_committed)  fputs(COL_GRN, stdout);
    else if (not_staged)  fputs(COL_RED, stdout);
    else if (untracked)  fputs(COL_YEL, stdout);

    fputs(thebranch, stdout);

    if (not_staged || to_be_committed || untracked) {
        fputs(COL_OFF, stdout);
    }
    if (aheadof_count > -1) {
        printf("%s%s%d%s) ", COL_RED, right_arrow, aheadof_count, COL_OFF);  // 27A1 + spc
    } else if (behind_count > -1) {
        printf("%s%s%d%s) ", COL_YEL, left_arrow, behind_count, COL_OFF);  // 2b05 is <-
    } else {
        printf(" %s) ", bullet);
    }
    return 0;
}


/**
 * Return true iff arg is equal to either short_sw or long_sw.
 */
bool isoneof(const char* short_sw, const char* long_sw, const char* arg) {
    return strcmp(short_sw, arg) * strcmp(long_sw, arg) == 0;
}


/**
 * Main routine - handle command line args and call appropriate sub-functions.
 */
int main(int argc, char **argv) {
    if (argc > 2) {
        fputs(MSG_WRONG_ARGS, stdout);
        return 2;
    }
    if (1 == argc) {
        return parse_input(false);  // no_unicode=false
    }
    // else: if (argc == 2):
    if (isoneof("-h", "--help", argv[1])) {
        fputs(MSG_USAGE, stdout);
        return 0;
    }
    if (isoneof("-v", "--version", argv[1])) {
        printf("ps1-4git v.%s\n", VERSION);
        return 0;
    }
    if (strcmp("--4bashrc", argv[1]) == 0) {
        fputs(PS1_SETUP_STRING, stdout);
        return 0;
    }
    if (isoneof("-n", "--no-unicode", argv[1])) {
        return parse_input(true);  // no_unicode=true

    } else {
        fputs(MSG_WRONG_ARGS, stdout);
        return 0;
    }
}


ps1/ps1-4git.c
/*
ps1-4git - A git-aware section for your PS1 prompt
Copyright (c) 2019, Yassen Damyanov

This program is free software: you can redistribute it and/or modify
it under the terms of the MIT License.
You should have received a copy of the MIT License along with this program.
If not, see <https://opensource.org/licenses/MIT>.
*/

/**
 * Scan `git status` output (reading it from stdin) to detect git branch
 * and status of workspace files, then output a PS1-suited prompt section
 * showing the branch and it's status.  See also README.md.
 */

#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>

#include "ps1-4git.h"

#define EOL 10
#define SPC 32
#define NUL '\0'
#define COMMA ','

#define PATH_MAX 255
#define LINE_MAX 1024
#define BRANCH_MAX 255

#define COL_RED "\001\033[31m\002"
#define COL_GRN "\001\033[32m\002"
#define COL_YEL "\001\033[33m\002"
#define COL_OFF "\001\033[0m\002"

#define PS1_SETUP_STRING "PS1=\"$PS1\\$(git status 2>&1 | ps1-4git)\"  # ps1-4git-MARKER: Do NOT remove marker\n"
#define MSG_WRONG_ARGS "Wrong arguments. Try 'ps1-4git -h'\n"

const char* MSG_USAGE =
    "ps1-4git - git-aware section for your PS1 prompt\n"
    "https://github.com/yassen-itlabs/git-goodies\n"
    "\n"
    "Usage:\n"
    "\n"
    "  ps1-4git -h|--help          Print this help screen.\n"
    "  ps1-4git -v|--version       Show program version.\n"
    "\n"
    "  git status 2>&1 | ps1-4git [-n|--no-unicode]\n"
    "                                Parse 'git status' output and print informative\n"
    "                                prompt ending string.\n"
    "\n"
    "                             -n|--no-unicode switch causes the output to be\n"
    "                                ascii only (no fancy arrows or bullets)\n"
    "\n"
    "                                Put this in $() at the end of your PS1 like:\n"
    "                                  PS1=\"$PS1\\$(git status 2>&1 | ps1-4git)\"\n"
    "\n"
    "  ps1-4git --4bashrc            Print a PS1 definition like the one above\n"
    "                                suitable for inclusion in .bashrc.\n"
;


/**
 * Return index of given char's occurance (searching left to right)
 * in str if found, -1 otherwise.
 */
int indexof(const char ch, const char* str, int start) {
    int idx = start;
    while (str[idx] != NUL && str[idx] != ch) {
        idx ++;
    }
    return (NUL == str[idx])? -1 : idx;
}


/**
 * Return index of given char's occurance (searching right to left)
 * in str if found, -1 otherwise.
 */
int rindexof(const char ch, const char* str, int start) {
    int idx = (start > -1)? start : (int)strlen(str) - 1;
    while (idx >= 0 && str[idx] != ch) {
        idx --;
    }
    return idx;
}


/**
 * Return true if string a starts with string b, false otherwise.
 */
bool startswith(const char* a, const char* b) {
    return (strncmp(a, b, strlen(b)) == 0);
}


/**
 * Expect the output of 'git status' on stdin; scan it for certain patterns
 * and output appropriate combination of strings with data forming he last
 * part of a PS1 shell prompt. Return appropriate result code.
 * If no_unicode is true -- use ascii characters in place of symbols (arrows,
 * bullets).
 */
int parse_input(bool no_unicode) {

    char line[LINE_MAX];
    char thebranch[BRANCH_MAX];

    bool to_be_committed = false;
    bool not_staged = false;
    bool untracked = false;

    // bool uptodate = false;
    int aheadof_count = -1;
    int behind_count = -1;

    while (fgets(line, sizeof(line), stdin) != NULL) {  // reads newline too
        if (startswith(line, "fatal: ")) {
            return 128;  // return same rc as 'git status' when not in a git repo

        } else if (startswith(line, "On branch ")) {
            // overwrite line EOL with NUL to get rid of it
            char* p = strchr(line, EOL);
            line[(int)(p - line)] = NUL;  // (p - line) is the index of the EOL char
            // get the last word as the branch name
            strcpy(thebranch, line + sizeof("On branch ") - 1);

        } else if (startswith(line, "Your branch is ahead of ")) {
            int right = rindexof(SPC, line, -1);
            int left = rindexof(SPC, line, right -1);
            aheadof_count = atoi(&line[left + 1]);

        } else if (startswith(line, "Your branch is behind ")) {
            int comma = rindexof(COMMA, line, -1);
            int right = rindexof(SPC, line, comma -1);
            int left = rindexof(SPC, line, right -1);
            behind_count = atoi(&line[left + 1]);

        // } else if (startswith(line, "Your branch is up to date ")) {
        //    uptodate = true;

        } else {
            not_staged = not_staged || startswith(line, "Changes not staged for commit:");
            to_be_committed = to_be_committed || startswith(line, "Changes to be committed:");
            untracked = untracked || startswith(line, "Untracked files:");
        }
    }

    // Unicodes: middle: A537 (used)  big: 27A1+space, super small bulet 2022 or 2981)
    // useful unicode char map: https://unicode-table.com/en/#hangul-syllables

    char* bullet = no_unicode? "*" : "ꔷ";
    char* left_arrow = no_unicode? " <" : "⬅ ";
    char* right_arrow = no_unicode? " >" : "➡ ";

    printf("(");  // TODO: denote push with ⭡2 (2B61), pull with ⭣3 (2B63)
    if (to_be_committed)  fputs(COL_GRN, stdout);
    else if (not_staged)  fputs(COL_RED, stdout);
    else if (untracked)  fputs(COL_YEL, stdout);

    fputs(thebranch, stdout);

    if (not_staged || to_be_committed || untracked) {
        fputs(COL_OFF, stdout);
    }
    if (aheadof_count > -1) {
        printf("%s%s%d%s) ", COL_RED, right_arrow, aheadof_count, COL_OFF);  // 27A1 + spc
    } else if (behind_count > -1) {
        printf("%s%s%d%s) ", COL_YEL, left_arrow, behind_count, COL_OFF);  // 2b05 is <-
    } else {
        printf(" %s) ", bullet);
    }
    return 0;
}


/**
 * Return true iff arg is equal to either short_sw or long_sw.
 */
bool isoneof(const char* short_sw, const char* long_sw, const char* arg) {
    return strcmp(short_sw, arg) * strcmp(long_sw, arg) == 0;
}


/**
 * Main routine - handle command line args and call appropriate sub-functions.
 */
int doit(int argc, char** argv) {
    if (argc > 2) {
        fputs(MSG_WRONG_ARGS, stdout);
        return 2;
    }
    if (1 == argc) {
        return parse_input(false);  // no_unicode=false
    }
    // else: if (argc == 2):
    if (isoneof("-h", "--help", argv[1])) {
        fputs(MSG_USAGE, stdout);
        return 0;
    }
    if (isoneof("-v", "--version", argv[1])) {
        printf("ps1-4git v.%s\n", VERSION);
        return 0;
    }
    if (strcmp("--4bashrc", argv[1]) == 0) {
        fputs(PS1_SETUP_STRING, stdout);
        return 0;
    }
    if (isoneof("-n", "--no-unicode", argv[1])) {
        return parse_input(true);  // no_unicode=true

    } else {
        fputs(MSG_WRONG_ARGS, stdout);
        return 0;
    }
}
